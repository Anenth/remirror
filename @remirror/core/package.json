{
  "name": "@remirror/core",
  "version": "0.11.0",
  "description": "Provides the core abstractions for managing your remirror editor functionality",
  "homepage": "https://github.com/remirror/remirror/tree/master/@remirror/core",
  "repository": "https://github.com/remirror/remirror/tree/master/@remirror/core",
  "license": "MIT",
  "author": "Ifiok Jr. <ifiokotung@gmail.com>",
  "sideEffects": false,
  "main": "lib/dist/index.js",
  "module": "lib/dist/core.esm.js",
  "browser": {
    "min-document": false
  },
  "types": "lib/index.d.ts",
  "files": [
    "internal",
    "lib",
    "src"
  ],
  "dependencies": {
    "@babel/runtime": "^7",
    "@remirror/core-constants": "^0.7.4",
    "@remirror/core-helpers": "^0.7.6",
    "@remirror/core-types": "^0.9.0",
    "@remirror/core-utils": "^0.8.0",
    "semver": "^7.3.0"
  },
  "peerDependencies": {
    "@remirror/pm": "^1.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "cjs": "lib/dist/core.cjs.js",
  "meta": {
    "sizeLimit": "70 KB"
  }
}
