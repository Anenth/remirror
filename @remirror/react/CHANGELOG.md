# @remirror/react

## 0.7.3
### Patch Changes

- 5f85c0de: Bump a new version to test out the changeset API.
- Updated dependencies [5f85c0de]
  - @remirror/core@0.7.3
  - @remirror/core-extensions@0.7.3
  - @remirror/react-portals@0.7.3
  - @remirror/react-ssr@0.7.3
  - @remirror/react-utils@0.7.3
  - @remirror/ui@0.7.3
