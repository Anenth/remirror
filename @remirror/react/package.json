{
  "name": "@remirror/react",
  "version": "0.13.1",
  "description": "react utilities for remirror",
  "homepage": "https://github.com/remirror/remirror/tree/next/@remirror/react",
  "repository": "https://github.com/remirror/remirror/tree/master/@remirror/react",
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "sideEffects": false,
  "main": "dist/react.cjs.js",
  "module": "dist/react.esm.js",
  "browser": {
    "./dist/react.cjs.js": "./dist/react.browser.cjs.js",
    "./dist/react.esm.js": "./dist/react.browser.esm.js"
  },
  "types": "dist/react.cjs.d.ts",
  "files": [
    "dist"
  ],
  "dependencies": {
    "@babel/runtime": "^7.10.3",
    "@remirror/extension-positioner": "^0.0.0",
    "@remirror/i18n": "^0.0.0",
    "@remirror/preset-core": "^0.0.0",
    "@remirror/preset-react": "^0.0.0",
    "@remirror/react-utils": "^0.7.6",
    "@types/hoist-non-react-statics": "^3.3.1",
    "@types/min-document": "^2.19.0",
    "@types/use-deep-compare-effect": "^1.2.0",
    "hoist-non-react-statics": "^3.3.2",
    "linaria": "^2.0.0-alpha.5",
    "min-document": "^2.19.0",
    "nanoevents": "^5.1.8",
    "nanoid": "^3.1.10",
    "react-use": "^15.3.0",
    "resize-observer-polyfill": "^1.5.1",
    "type-fest": "^0.15.1",
    "use-deep-compare-effect": "^1.3.1"
  },
  "devDependencies": {
    "@remirror/core": "^0.11.0",
    "@remirror/pm": "^0.0.0",
    "@testing-library/react": "^10.3.0",
    "@testing-library/react-hooks": "^3.3.0",
    "@types/react": "^16.9.38",
    "@types/react-dom": "^16.9.8",
    "@types/react-test-renderer": "^16.9.2",
    "@types/testing-library__react": "^10.0.1",
    "@types/testing-library__react-hooks": "^3.2.0",
    "jest-axe": "^3.4.0",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-test-renderer": "^16.13.1"
  },
  "peerDependencies": {
    "@remirror/core": "^0.11.0",
    "@remirror/pm": "^0.0.0",
    "@types/react": "^16.8.0",
    "@types/react-dom": "^16.8.0",
    "react": "^16.8.0",
    "react-dom": "^16.8.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "meta": {
    "sizeLimit": "30 KB"
  }
}
