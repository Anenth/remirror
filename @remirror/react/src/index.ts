export * from './node-views';
export * from './components';
export * from './hooks';
export * from './hocs';
export * from '@remirror/react-utils';
export * from './helpers';
