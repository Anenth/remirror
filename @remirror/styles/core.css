/**
 * Styles extracted from: @remirror/core/src/styles.ts
 */
.remirror-editor.ProseMirror {
  position: relative;
  word-wrap: break-word;
  white-space: pre-wrap;
  white-space: break-spaces;
  -webkit-font-variant-ligatures: none;
  font-variant-ligatures: none;
  font-feature-settings: 'liga' 0;
}
.remirror-editor.ProseMirror pre {
  white-space: pre-wrap;
}
.remirror-editor.ProseMirror li {
  position: relative;
}
.remirror-editor.ProseMirror[contenteditable='false'] {
  white-space: normal;
}
.remirror-editor.ProseMirror[contenteditable='true'] {
  white-space: pre-wrap;
}
.remirror-editor .ProseMirror-hideselection {
  caret-color: transparent;
}
.remirror-editor .ProseMirror-hideselection *::selection,
.remirror-editor .ProseMirror-hideselection *::-moz-selection {
  background: transparent;
}
.remirror-editor .ProseMirror-selectednode {
  outline: 2px solid #8cf;
}
.remirror-editor li.ProseMirror-selectednode {
  outline: none;
}
.remirror-editor li.ProseMirror-selectednode:after {
  content: '';
  position: absolute;
  left: -32px;
  right: -2px;
  top: -2px;
  bottom: -2px;
  border: 2px solid #8cf;
  pointer-events: none;
}
