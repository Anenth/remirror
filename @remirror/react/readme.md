# @remirror/react

[![npm bundle size (scoped)](https://img.shields.io/bundlephobia/minzip/@remirror/react.svg?)](https://bundlephobia.com/result?p=@remirror/react)
[![npm](https://img.shields.io/npm/dm/@remirror/react.svg?&logo=npm)](https://www.npmjs.com/package/@remirror/react)

The react components for the remirror editor

## Installation

```bash
yarn add @remirror/react # yarn
pnpm add @remirror/react # pnpm
npm install @remirror/react # npm
```

## Usage

For in depth usage with proper code example see the [docs](https://docs.remirror.org).

### Controlled Editor

The following example converts all the content to text and appends a list item to the end of every
editor after every edit.

Don't do this, as it would actually be a terrible user experience. But it shows what can be done. A
more meaningful example will be created soon.

```ts
import React, { useCallback } from 'react';
import { BoldExtension } from 'remirror/extension/bold';
import { ListPreset } from 'remirror/preset/list';
import { fromHtml, RemirrorEventListener } from 'remirror/core';
import {
  RemirrorProvider,
  createReactManager,
  ReactCombinedUnion,
  useManager,
  useCreateExtension,
  useCreatePreset,
} from 'remirror/react';

type Combined = ReactCombinedUnion<ListPreset | BoldExtension>;

const EditorWrapper = () => {
  const boldExtension = useCreateExtension(BoldExtension);
  const listPreset = useCreatePreset(ListPreset);
  const manager = useManager<Combined>([boldExtension, listPreset]);

  const initialValue = manager.createState({
    content: '<p>This is the initial value</p>',
    stringHandler: fromHtml,
  });

  const [value, setValue] = useState(initialValue);

  const onChange: RemirrorEventListener<Combined> = useCallback(
    ({ getText, createStateFromContent }) => {
      const newValue = createStateFromContent(`${getText()}<ul><li>Surprise!!!</li></ul>`);
      setValue(newValue);
    },
    [],
  );

  return (
    <RemirrorProvider manager={manager} stringHandler={fromHtml} value={value} onChange={onChange}>
      <div />
    </RemirrorProvider>
  );
};
```
