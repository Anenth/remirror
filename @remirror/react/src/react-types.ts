import { Interpolation, ObjectInterpolation } from '@emotion/core';
import { ReactNode, Ref } from 'react';

import {
  AnyExtension,
  ChainedFromExtensions,
  CommandsFromExtensions,
  CompareStateParameter,
  EditorManager,
  EditorSchema,
  EditorState,
  EditorStateParameter,
  EditorView,
  EditorViewParameter,
  ElementParameter,
  FromToParameter,
  GetExtensionUnion,
  ObjectNode,
  PlainObject,
  Position,
  PositionParameter,
  ProsemirrorNode,
  RemirrorContentType,
  RemirrorInterpolation,
  RenderEnvironment,
  SchemaFromExtension,
  StringHandlerParameter,
  TextParameter,
  Transaction,
  TransactionParameter,
  TransactionTransformer,
} from '@remirror/core';

/**
 * The type of arguments acceptable for the focus parameter.
 *
 * - Can be a selection of `{ from, to }
 * - A single position with a `number`
 * - `start` | `end`
 * - `true` which sets the focus to the current position or start.
 */
export type FocusType = FromToParameter | number | 'start' | 'end' | boolean;

export interface RenderEditorProps<ManagerType extends EditorManager = EditorManager>
  extends StringHandlerParameter {
  /**
   * Pass in the extension manager.
   *
   * The manager is responsible for handling all Prosemirror related
   * functionality.
   */
  manager: ManagerType;

  /**
   * Set the starting value object of the editor.
   *
   * Without setting onStateChange remirror renders as an uncontrolled
   * component. Value changes are passed back out of the editor and there is now
   * way to set the value via props. As a result this is the only opportunity to
   * directly control the rendered text.
   *
   * @defaultValue `{ type: 'doc', content: [{ type: 'paragraph' }] }`
   */
  initialContent: RemirrorContentType;

  /**
   * If this exists the editor becomes a controlled component. Nothing will be
   * updated unless you explicitly set the value prop to the updated state.
   *
   * Without a deep understanding of Prosemirror this is not recommended.
   */
  onStateChange?: (params: RemirrorStateListenerParameter<GetExtensionUnion<ManagerType>>) => void;

  /**
   * When onStateChange is defined this prop is used to set the next state value
   * of the remirror editor.
   */
  value?: EditorState<SchemaFromExtension<GetExtensionUnion<ManagerType>>> | null;

  /**
   * Adds attributes directly to the prosemirror html element.
   *
   * @defaultValue `{}`
   */
  attributes: Record<string, string> | AttributePropFunction;

  /**
   * Determines whether this editor is editable or not.
   *
   * @defaultValue true
   */
  editable: boolean;

  /**
   * When set to true focus will be place on the editor as soon as it first
   * loads.
   *
   * @defaultValue false
   */
  autoFocus?: FocusType;

  /**
   * An event listener which is called whenever the editor gains focus.
   */
  onFocus?: (
    params: RemirrorEventListenerParameter<GetExtensionUnion<ManagerType>>,
    event: Event,
  ) => void;

  /**
   * An event listener which is called whenever the editor is blurred.
   */
  onBlur?: (
    params: RemirrorEventListenerParameter<GetExtensionUnion<ManagerType>>,
    event: Event,
  ) => void;

  /**
   * Called on the first render when the prosemirror instance first becomes
   * available
   */
  onFirstRender?: RemirrorEventListener<GetExtensionUnion<ManagerType>>;

  /**
   * Called on every change to the Prosemirror state.
   */
  onChange?: RemirrorEventListener<GetExtensionUnion<ManagerType>>;

  /**
   * The render prop that takes the injected remirror params and returns an
   * element to render. The editor view is automatically attached to the DOM.
   */
  children: RenderPropFunction<ManagerType>;

  /**
   * A method called when the editor is dispatching the transaction.
   *
   * @remarks
   * Use this to update the transaction which will be used to update the editor
   * state.
   */
  onDispatchTransaction: TransactionTransformer<
    SchemaFromExtension<GetExtensionUnion<ManagerType>>
  >;

  /**
   * Sets the accessibility label for the editor instance.
   *
   * @defaultValue ''
   */
  label: string;

  /**
   * Determines whether or not to use the built in extensions.
   *
   * @remarks
   *
   * Use this if you would like to take full control of all your extensions and
   * have some understanding of the underlying Prosemirror internals.
   *
   * ```ts
   * const builtInExtensions = [new DocExtension(), new TextExtension(), new ParagraphExtension()]
   * ```
   *
   * @defaultValue true
   */
  usesBuiltInExtensions: boolean;

  /**
   * Determine whether the editor should use default styles.
   *
   * @defaultValue true
   */
  usesDefaultStyles: boolean;

  /**
   * Additional editor styles passed into prosemirror. Used to provide styles
   * for the text, nodes and marks rendered in the editor.
   *
   * @defaultValue `{}`
   *
   * @deprecated use `styles` or `css` prop instead.
   */
  editorStyles: RemirrorInterpolation;

  /**
   * Addition styles that will be passed directly to the prosemirror editor dom node.
   *
   * This can be used to provide extra styles
   *
   */
  styles?: RemirrorInterpolation;

  /**
   * Emotion css prop for injecting styles into a component.
   */
  css?: Interpolation;

  /**
   * Determine whether the Prosemirror view is inserted at the `start` or `end`
   * of it's container DOM element.
   *
   * @defaultValue 'end'
   */
  insertPosition: 'start' | 'end';

  /**
   * By default remirror will work out whether this is a dom environment or
   * server environment for SSR rendering. You can override this behaviour here
   * when required.
   */
  forceEnvironment?: RenderEnvironment;

  /**
   * Set to true to ignore the hydration warning for a mismatch between the
   * rendered server and client content.
   *
   * @remarks
   *
   * This is a potential solution for those who require server side rendering.
   *
   * While on the server the prosemirror document is transformed into a react
   * component so that it can be rendered. The moment it enters the DOM
   * environment prosemirror takes over control of the root element. The problem
   * is that this will always see this hydration warning on the client:
   *
   * `Warning: Did not expect server HTML to contain a <div> in <div>.`
   *
   * Setting this to true removes the warning at the cost of a slightly slower
   * start up time. It uses the two pass solution mentioned in the react docs.
   * See {@link https://reactjs.org/docs/react-dom.html#hydrate}.
   *
   * For ease of use this prop copies the name used by react for DOM Elements.
   * See {@link
   * https://reactjs.org/docs/dom-elements.html#suppresshydrationwarning.
   */
  suppressHydrationWarning?: boolean;

  /**
   * The value to use for empty content.
   *
   * This is the value used for an empty editor or when `resetContent` is called.
   *
   * @defaultValue EMPTY_PARAGRAPH_NODE
   */
  fallbackContent: ObjectNode | ProsemirrorNode;
}

export interface Positioner<ExtensionUnion extends AnyExtension = any> {
  /**
   * The default and initial position value. This is used at the start and
   * whenever isActive becomes false
   */
  initialPosition: Position;

  /**
   * Determines whether anything has changed and whether to continue with a
   * recalculation. By default this is only true when the document has changed.
   *
   * @remarks
   *
   * Sometimes it is useful to recalculate the positioner on every state update. In this case
   * you can set this method to always return true.
   *
   * ```ts
   * const positioner: Positioner = {
   *   hasChanged: () => true
   *
   * };
   *
   * @param params
   */
  hasChanged: (params: CompareStateParameter<SchemaFromExtension<ExtensionUnion>>) => boolean;

  /**
   * Determines whether the positioner should be active
   */
  isActive: (params: GetPositionParameter<ExtensionUnion>) => boolean;

  /**
   * Calculate and return a new position (only called when `hasChanged` and
   * `isActive` return true)
   */
  getPosition: (params: GetPositionParameter<ExtensionUnion>) => Position;
}

export type CalculatePositionerParameter<
  ExtensionUnion extends AnyExtension = any
> = PositionerIdParameter & Positioner<ExtensionUnion>;

export type GetPositionerPropsConfig<
  ExtensionUnion extends AnyExtension = any,
  GRefKey extends string = 'ref'
> = RefParameter<GRefKey> & Partial<Positioner<ExtensionUnion>> & PositionerIdParameter;

export interface RefParameter<GRefKey = 'ref'> {
  /**
   * A custom ref key which allows a reference to be obtained from non standard
   * components.
   *
   * @defaultValue 'ref'
   */
  refKey?: GRefKey;
}

export type PositionerProps = IsActiveParameter & Position;

export interface GetRootPropsConfig<GRefKey extends string = 'ref'>
  extends RefParameter<GRefKey>,
    PlainObject {
  editorStyles?: Interpolation;
}

export type RefKeyRootProps<GRefKey extends string = 'ref'> = {
  [P in Exclude<GRefKey, 'key'>]: Ref<any>;
} & {
  css: Interpolation | ((theme: any) => Interpolation);
  key: string;
  children: ReactNode;
} & PlainObject;

export type GetPositionerReturn<GRefKey extends string = 'ref'> =
  // GRefKey extends 'ref'
  //   ? PositionerProps & { ref: Ref<any> }
  // :
  PositionerProps & { [P in GRefKey]: Ref<any> };

/**
 * These are the props passed to the render function provided when setting up
 * your editor.
 */
export interface InjectedRenderEditorProps<ManagerType extends EditorManager = any> {
  /**
   * An instance of the extension manager
   */
  manager: ManagerType;
  /**
   * The prosemirror view
   */
  view: EditorView<SchemaFromExtension<GetExtensionUnion<ManagerType>>>;

  /**
   * A map of all actions made available by the configured extensions.
   */
  commands: CommandsFromExtensions<GetExtensionUnion<ManagerType>>;

  /**
   * The chainable commands.
   */
  chain: ChainedFromExtensions<GetExtensionUnion<ManagerType>>;

  /**
   * The unique id for the editor instance.
   */
  uid: string;

  /**
   * Clears all editor content
   *
   * @param triggerOnChange - whether onChange handlers should be triggered by
   * the update.
   */
  clearContent: (triggerOnChange?: boolean) => void;

  /**
   * Replace all editor content with the new content.
   *
   * @remarks
   *
   * Allows for the editor content to be overridden by force.
   *
   * @param triggerOnChange - whether onChange handlers should be triggered by
   * the update.
   */
  setContent: (content: RemirrorContentType, triggerOnChange?: boolean) => void;

  /**
   * A function that returns props which should be spread on a react element and
   * declare it as the editor root (where the editor is injected in the DOM).
   *
   * @remarks
   * By default remirror will add the prosemirror editor instance directly into
   * the first child element it receives. Using this method gives you full
   * control over where the editor should be injected.
   *
   * **IMPORTANT** For this to work properly you will need to use the JSX pragma from
   * `emotion/core` so that the `css` prop can properly be added to the element.
   *
   * **IMPORTANT** In order to support SSR pre-rendering this should only be spread
   * on a component with NO children.
   *
   * **Example with indirectly nested components**
   *
   * ```tsx
   * // @jsx jsx
   * import { jsx } from '@emotion/core';
   * import { ManagedRemirrorProvider, RemirrorManager, useRemirrorContext } from '@remirror/react';
   *
   * const InnerEditor = () => {
   *   const { getRootProps } = useRemirrorContext();
   *   return <div {...getRootProps()} />;
   * }
   *
   * const Editor = () => {
   *   return (
   *     <RemirrorManager>
   *       <ManagedRemirrorProvider>
   *         <InnerEditor />
   *       </ManagedRemirrorProvider>
   *     </RemirrorManager>
   *   );
   * }
   * ```
   */
  getRootProps: <GRefKey extends string = 'ref'>(
    options?: GetRootPropsConfig<GRefKey>,
  ) => RefKeyRootProps<GRefKey>;

  /**
   * Attach these props to a component to inject it with position data.
   * Typically this is used for creating menu components.
   *
   * A custom positioner can be passed in to update the method used to calculate
   * the position.
   */
  getPositionerProps: <GRefKey extends string = 'ref'>(
    options: GetPositionerPropsConfig<GetExtensionUnion<ManagerType>, GRefKey>,
  ) => GetPositionerReturn<GRefKey>;

  /**
   * The previous and next state
   */
  state: CompareStateParameter<SchemaFromExtension<GetExtensionUnion<ManagerType>>>;

  /**
   * Focus the editor at the `start` | `end` a specific position or at a valid range between `{ from, to }`
   */
  focus: (position?: FocusType) => void;
}

/**
 * A function that takes the injected remirror params and returns JSX to render.
 *
 * @param - injected remirror params
 */
export type RenderPropFunction<ManagerType extends EditorManager = any> = (
  params: InjectedRenderEditorProps<ManagerType>,
) => JSX.Element;

export interface RemirrorGetterParameter {
  /**
   * Get the current HTML from the latest editor state.
   */
  getHTML: () => string;

  /**
   * Get the current raw text from the latest editor state.
   *
   * @param lineBreakDivider - the divider to use for new lines defaults to
   * '\n\n'
   */
  getText: (lineBreakDivider?: string) => string;

  /**
   * Get the full JSON representation of the state (including the selection
   * information)
   */
  getJSON: () => ObjectNode;

  /**
   * Get a representation of the editor content as an ObjectNode which can be
   * used to set content for and editor.
   */
  getObjectNode: () => ObjectNode;
}

export interface BaseListenerParameter<ExtensionUnion extends AnyExtension = any>
  extends EditorViewParameter<SchemaFromExtension<ExtensionUnion>>,
    RemirrorGetterParameter {
  /**
   * The original transaction which caused this state update.
   *
   * This allows for inspecting the reason behind the state change.
   * When undefined this means that the state was updated externally.
   *
   * If available:
   * - Metadata on the transaction can be inspected. `tr.getMeta`
   * - Was the change caused by added / removed content? `tr.docChanged`
   * - Was ths change caused by an updated selection? `tr.selectionSet`
   * - `tr.steps` can be inspected for further granularity.
   */
  tr?: Transaction<SchemaFromExtension<ExtensionUnion>>;

  /**
   * A shorthand way of checking whether the update was triggered by editor usage (internal) or
   * overwriting the state.
   *
   * - `true` The update was triggered by a change in the prosemirror doc or an update to the selection.
   * In these cases `tr` will have a value.
   * - `false` The update was caused by a call to `setContent` or `resetContent`
   */
  internalUpdate: boolean;
}

export interface RemirrorEventListenerParameter<ExtensionUnion extends AnyExtension = any>
  extends EditorStateParameter<SchemaFromExtension<ExtensionUnion>>,
    BaseListenerParameter<ExtensionUnion> {}

export interface RemirrorStateListenerParameter<ExtensionUnion extends AnyExtension = any>
  extends CompareStateParameter<SchemaFromExtension<ExtensionUnion>>,
    BaseListenerParameter<ExtensionUnion> {
  /**
   * Manually create a new state object with the desired content.
   */
  createStateFromContent: (
    content: RemirrorContentType,
  ) => EditorState<SchemaFromExtension<ExtensionUnion>>;
}

export type RemirrorEventListener<ExtensionUnion extends AnyExtension = any> = (
  params: RemirrorEventListenerParameter<ExtensionUnion>,
) => void;

export type AttributePropFunction<ExtensionUnion extends AnyExtension = any> = (
  params: RemirrorEventListenerParameter<ExtensionUnion>,
) => Record<string, string>;

export interface PlaceholderConfig extends TextParameter {
  className: string;
  style: ObjectInterpolation<undefined>;
}

export type PositionerMapValue = ElementParameter & {
  prev: PositionerProps;
};

export interface PositionerRefFactoryParameter extends PositionerIdParameter, PositionParameter {}

export interface GetPositionParameter<ExtensionUnion extends AnyExtension = any>
  extends EditorViewParameter<SchemaFromExtension<ExtensionUnion>>,
    ElementParameter,
    CompareStateParameter<SchemaFromExtension<ExtensionUnion>> {}

export interface PositionerIdParameter {
  /**
   * A unique id for the positioner.
   *
   * This is used to map the ref of the tracked component to a unique id and
   * cant be updated without losing track of the component's reference element.
   */
  positionerId: string;
}

export interface IsActiveParameter {
  /**
   * A boolean value determining whether the positioner should be active.
   */
  isActive: boolean;
}

export interface PositionerParameter {
  /**
   * The positioner object which determines how the changes in the view impact
   * the calculated position.
   */
  positioner: Partial<Positioner>;
}

export interface UsePositionerParameter<GRefKey extends string = 'ref'>
  extends PositionerIdParameter,
    PositionerParameter,
    RefParameter<GRefKey> {}

export interface UpdateStateParameter<GSchema extends EditorSchema = any>
  extends Partial<TransactionParameter<GSchema>>,
    EditorStateParameter<GSchema> {
  /**
   * Whether or not to trigger this as a change and call any handlers.
   *
   * @defaultValue true
   */
  triggerOnChange?: boolean;

  /**
   * Called after the state has updated.
   */
  onUpdate?: () => void;
}

export interface EditorStateEventListenerParameter<
  ExtensionUnion extends AnyExtension = any,
  GSchema extends EditorSchema = any
>
  extends Partial<CompareStateParameter<GSchema>>,
    Pick<BaseListenerParameter<ExtensionUnion>, 'tr'> {}

export interface RemirrorState<GSchema extends EditorSchema = any> {
  /**
   * The Prosemirror editor state
   */
  editor: CompareStateParameter<GSchema>;
  /**
   * Used when suppressHydrationWarning is true to determine when it's okay to
   * render the client content.
   */
  shouldRenderClient?: boolean;
}

export interface ListenerParameter<
  ExtensionUnion extends AnyExtension = any,
  GSchema extends EditorSchema = any
>
  extends Partial<EditorStateParameter<GSchema>>,
    Pick<BaseListenerParameter<ExtensionUnion>, 'tr'> {}
