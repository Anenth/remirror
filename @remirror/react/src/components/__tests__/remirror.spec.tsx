import { fireEvent, render } from '@testing-library/react';
import { axe } from 'jest-axe';
import React from 'react';
import { renderToString } from 'react-dom/server';

import { fromHTML } from '@remirror/core';
import { createTestManager } from '@remirror/test-fixtures';

import { RenderEditor } from '..';
import { InjectedRenderEditorProps } from '../../react-types';

const textContent = `This is editor text`;
const label = 'Remirror editor';
const handlers = {
  onChange: jest.fn(),
  onBlur: jest.fn(),
  onFocus: jest.fn(),
  onFirstRender: jest.fn(),
};

test('should be called via a render prop', () => {
  const mock = jest.fn(() => <div />);
  const { getByLabelText } = render(
    <RenderEditor manager={createTestManager()} label={label} {...handlers}>
      {mock}
    </RenderEditor>,
  );

  expect(mock).toHaveBeenCalledWith(expect.any(Object));
  expect(handlers.onFirstRender).toHaveBeenCalledWith(expect.any(Object));
  expect(handlers.onFirstRender.mock.calls[0][0].getText()).toBe('');
  expect(handlers.onFirstRender.mock.calls[0][0].getJSON().doc.type).toBe('doc');
  expect(handlers.onFirstRender.mock.calls[0][0].getHTML().type).toBe(undefined);

  const editorNode = getByLabelText(label);

  expect(editorNode).toHaveAttribute('role', 'textbox');
});

test('can suppressHydrationWarning without breaking', () => {
  const mock = jest.fn(() => <div />);
  const { getByLabelText } = render(
    <RenderEditor
      manager={createTestManager()}
      label={label}
      {...handlers}
      suppressHydrationWarning={true}
    >
      {mock}
    </RenderEditor>,
  );

  expect(mock).toHaveBeenCalledWith(expect.any(Object));
  expect(handlers.onFirstRender).toHaveBeenCalledWith(expect.any(Object));
  expect(handlers.onFirstRender.mock.calls[0][0].getText()).toBe('');
  expect(handlers.onFirstRender.mock.calls[0][0].getJSON().doc.type).toBe('doc');
  expect(handlers.onFirstRender.mock.calls[0][0].getHTML().type).toBe(undefined);

  const editorNode = getByLabelText(label);

  expect(editorNode).toHaveAttribute('role', 'textbox');
});

describe('basic functionality', () => {
  it('is accessible', async () => {
    const results = await axe(
      renderToString(<RenderEditor manager={createTestManager()}>{() => <div />}</RenderEditor>),
    );

    expect(results).toHaveNoViolations();
  });

  it("doesn't render the editor without `children` as a render prop", () => {
    const spy = jest.spyOn(console, 'error').mockImplementation(() => undefined);

    expect(() =>
      // @ts-expect-error
      render(<RenderEditor label={label} manager={createTestManager()} />),
    ).toThrowErrorMatchingInlineSnapshot(
      `"The child argument to the Remirror component must be a function."`,
    );
    expect(spy).toHaveBeenCalledTimes(2);

    spy.mockRestore();
  });

  it('should allow text input and fire all handlers', () => {
    let setContent: InjectedRenderEditorProps['setContent'] = jest.fn();
    const mock = jest.fn((value: InjectedRenderEditorProps) => {
      setContent = value.setContent;
      return <div />;
    });
    const { getByLabelText } = render(
      <RenderEditor
        label={label}
        {...handlers}
        manager={createTestManager()}
        stringHandler={fromHTML}
      >
        {mock}
      </RenderEditor>,
    );

    setContent(`<p>${textContent}</p>`, true);
    const editorNode = getByLabelText(label);

    expect(handlers.onChange).toHaveBeenCalledTimes(1);
    expect(handlers.onFirstRender.mock.calls[0][0].getText()).toBe(textContent);

    fireEvent.blur(editorNode);

    expect(handlers.onBlur).toHaveBeenCalledTimes(1);

    fireEvent.focus(editorNode);

    expect(handlers.onFocus).toHaveBeenCalledTimes(1);
  });

  it('changes when the editable prop changes', () => {
    const mock = jest.fn(() => <div />);
    const manager = createTestManager();
    const El = ({ editable }: { editable: boolean }) => (
      <RenderEditor editable={editable} label={label} manager={manager}>
        {mock}
      </RenderEditor>
    );
    const { rerender, getByLabelText } = render(<El editable={true} />);

    expect(getByLabelText(label)).toHaveAttribute('contenteditable', 'true');

    rerender(<El editable={false} />);

    expect(getByLabelText(label)).toHaveAttribute('contenteditable', 'false');
  });
});

describe('initialContent', () => {
  it('should render with string content', () => {
    const { container } = render(
      <RenderEditor
        label={label}
        {...handlers}
        manager={createTestManager()}
        initialContent={'<p>Hello</p>'}
        stringHandler={fromHTML}
      >
        {() => <div />}
      </RenderEditor>,
    );

    expect(container.innerHTML).toInclude('Hello');
  });

  it('renders with json', () => {
    const content = {
      type: 'doc',
      content: [{ type: 'paragraph', content: [{ type: 'text', text: 'Hello' }] }],
    };

    const { container } = render(
      <RenderEditor
        label={label}
        {...handlers}
        manager={createTestManager()}
        initialContent={content}
      >
        {() => <div />}
      </RenderEditor>,
    );

    expect(container.innerHTML).toInclude('Hello');
  });
});

describe('focus', () => {
  const content = {
    type: 'doc',
    content: [{ type: 'paragraph', content: [{ type: 'text', text: 'A sentence here' }] }],
  };

  let props!: InjectedRenderEditorProps<any>;
  let editorNode: HTMLElement;

  beforeEach(() => {
    jest.useFakeTimers();
    const { getByRole } = render(
      <RenderEditor
        label={label}
        {...handlers}
        manager={createTestManager()}
        initialContent={content}
        autoFocus={true}
      >
        {(context) => {
          props = context;
          return <div />;
        }}
      </RenderEditor>,
    );

    editorNode = getByRole('textbox');
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should do nothing when focusing on a focused editor without a new position', () => {
    expect(props.state.newState.selection.from).toBe(1);

    props.focus();
    jest.runAllTimers();

    expect(props.state.newState.selection.from).toBe(1);
    // TODO expect(props.view.hasFocus()).toBeTrue();
  });

  it('can focus on the end', () => {
    fireEvent.blur(editorNode);
    props.focus('end');
    jest.runAllTimers();

    expect(props.state.newState.selection.from).toBe(16);
  });

  it('can focus on the start even when already focused', () => {
    props.focus('start');
    jest.runAllTimers();

    expect(props.state.newState.selection.from).toBe(1);
  });

  it('can specify the exact position for the blurred editor', () => {
    fireEvent.blur(editorNode);
    props.focus(10);
    jest.runAllTimers();

    expect(props.state.newState.selection.from).toBe(10);
  });

  const expected = { from: 2, to: 5 };

  it('can specify the selection for the editor', () => {
    fireEvent.blur(editorNode);
    props.focus(expected);
    jest.runAllTimers();
    {
      const { from, to } = props.state.newState.selection;

      expect({ from, to }).toEqual(expected);
    }
  });

  it('restores the previous selection when focused without a parameter', () => {
    fireEvent.blur(editorNode);
    props.focus(expected);
    jest.runAllTimers();
    {
      const { from, to } = props.state.newState.selection;

      expect({ from, to }).toEqual(expected);
    }

    fireEvent.blur(editorNode);
    props.focus();
    jest.runAllTimers();
    {
      const { from, to } = props.state.newState.selection;

      expect({ from, to }).toEqual(expected);
      // expect(props.view.hasFocus()).toBeTrue();
    }
  });

  it('should do nothing when passing `false`', () => {
    fireEvent.blur(editorNode);
    props.focus(false);
    jest.runAllTimers();

    expect(props.state.newState.selection.from).toBe(1);
  });
});
