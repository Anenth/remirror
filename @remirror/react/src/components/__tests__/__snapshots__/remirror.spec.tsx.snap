// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remirror Controlled Component should call onStateChange 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "align": null,
        "color": null,
        "id": null,
        "indent": 0,
        "lineSpacing": null,
      },
      "content": Array [
        Object {
          "text": "Hello",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;
